core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsTemplateEngine.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.statement;
+
+import java.util.HashSet;
+import java.util.Set;
+
+class DefineNamedBindingsTemplateEngine implements TemplateEngine {
+    private final DefineNamedBindingMode mode;
+    private final TemplateEngine delegate;
+
+    public DefineNamedBindingsTemplateEngine(DefineNamedBindingMode mode, TemplateEngine delegate) {
+        this.mode = mode;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public String render(String template, StatementContext ctx) {
+        final Set<String> binds = new HashSet<>(ctx.getBinding().getNames());
+        binds.removeAll(ctx.getAttributes().keySet());
+        for (String boundButNotDefined : binds) {
+            ctx.getBinding().findForName(boundButNotDefined, ctx)
+                .flatMap(mode::apply)
```

Comment:
I'm not seeing the point of the Optional here. We call `ctx.getBinding().getNames()` to get the name, so `ctx.getBinding().findForName` will never be able to return `empty()`, will it? I think it should be `orElseThrow` because we expect a value always.

Reply:
It's legitimate to implement a `NamedArgumentFinder` that has a name but is unable to find a value.  I don't see why you would do that, but if you do, this method should ignore it rather than exploding I think.

Reply:
If you're certain an empty return can be legit, then ok. My concern was the functional expression of allowing empty values to go unnoticed, not so much whether they should never be empty or not.

Reply:
I'm not certain it's legit.  But it doesn't feel like this is the right place to start enforcing a new policy like that, and not mixed into an otherwise unrelated PR.

Reply:
If the binding is truly accidentally forgotten, it should blow up later during binding.